functions/database/database.js

```diff
@@ -0,0 +1,316 @@
+const RELEASE_STATES = require("../utils/releaseStates.js");
+const {
+  validateNewReleaseStructure,
+  validateRelease,
+} = require("../validation/validation.js");
+const {parseCommitTitleFromMessage} = require("../utils/utils.js");
+const {REPO_URL} = require("../github/github.js");
+
+/**
+ * Retrieve the release ID of the Firestore release document
+ * with a given release name.
+ *
+ * @param {string} releaseName The release name of the release
+ * document to fetch.
+ * @throws {Error} If there is not exactly one release with the given name.
+ * @return {Promise<string>} A promise that resolves to the release ID.
+ */
+async function getReleaseID(releaseName) {
+  const releaseSnapshot = await db.collection("releases")
+      .where("releaseName", "==", releaseName)
+      .get();
+
+  if (releaseSnapshot.empty) {
+    throw new Error(`No release found with name: ${releaseName}`);
+  }
+
+  // There should only be one release with a given name
+  if (releaseSnapshot.size > 1) {
+    throw new Error(`${releaseSnapshot.size} releases found with name:
+     ${releaseName}`);
+  }
+
+  const releaseId = releaseSnapshot.docs[0].id;
+
+  return releaseId;
+}
+
+/**
+ * Retrieves release data from the database given a release ID.
+ *
+ * @param {string} releaseId The ID of the release to fetch.
+ * @return {Promise<Object>} A promise that resolves to the release data.
+ */
+async function getReleaseData(releaseId) {
+  const releaseSnapshot = await db.collection("releases")
+      .doc(releaseId)
+      .get();
+
+  if (!releaseSnapshot.exists) {
+    throw new Error(`No release found with ID: ${releaseId}`);
+  }
+
+  const releaseData = releaseSnapshot.data();
+
+  return releaseData;
+}
+
+/** Retrieve all past releases from Firestore. This includes releases that are
+ * currently active.
+ *
+ * @return {Promise<Array>} Returns a promise that resolves to an array
+ * of Firestore document data of past releases.
+ * @throws {Error} If there is an error in fetching data from Firestore,
+ * the promise is rejected with an error.
+ */
+async function getPreviousReleaseData() {
+  const today = new Date();
+  const snapshot = await db.collection("releases")
+      .where("releaseDate", "<", today)
+      .get();
+
+  const pastReleases = snapshot.docs.map((doc) => doc.data());
+
+  return pastReleases;
+}
+
+/** Write new releases to Firestore. This function assumes that the contents
+ * of the releases are validated. newReleases is an array of release objects
+ * that have the following structure:
+ * {
+ *   releaseName: string,
+ *   releaseOperator: string,
+ *   codeFreezeDate: Date,
+ *   releaseDate: Date,
+ * }
+ *
+ * @param {Object} newReleases - Releases to store in Firestore
+ */
+async function addReleases(newReleases) {
+  // Validate the structure of the new releases before we store
+  validateNewReleaseStructure(newReleases);
+
+  const batch = db.batch();
+
+  newReleases.forEach((release) => {
```

Comment:
I would much rather have a `addRelease` function that this maps off to, but I recognize the desire to retain atomic transactions. An alternative would be a `releaseToFirestoreObject` method (or whatever name you think suites it- but you get the idea) that this maps newReleases to, and then creates a batch for:

```js
async function addReleases(newReleases) {
  validateNewReleaseStructure(newReleases);

  const newReleaseObjects = newReleases.map(release => releaseToFirestoreObject(release))

  const batch = db.batch();

  newReleaseObjects.forEach((release) => {
    const newReleaseRef = db.collection("releases").doc();

    batch.set(newReleaseRef, release);
  });

  await batch.commit();
}
```

Alternatively, this would be fine too:

```js
async function addReleases(newReleases) {
  validateNewReleaseStructure(newReleases);

  const batch = db.batch();

  newReleases.forEach((release) => {
    const newReleaseRef = db.collection("releases").doc();
    const releaseData = releaseToReleaseData(release);

    batch.set(newReleaseRef, releaseData);
  });

  await batch.commit();
}
```

Reply:
Yeah I agree, these functions could have been more modular. I refactored all functions that have batched writes here https://github.com/firebase/firebase-release-dashboard/pull/8/commits/030399c9ed011642c29c1d0d1c946d65c8e3d17f.